docs/proposals-accepted/202301-distributed-query-execution.md

```diff
@@ -0,0 +1,159 @@
+which can be transformed as
+
+```
+sum(
+  coalesce(
+    sum(rate(metric[2m])),
+    sum(rate(metric[2m]))
+  )
+)
+```
+
+### How
+
+The proposed method of transforming the query is extending the Thanos Engine with a logical optimizer that has
+references to other query engines. An example API could look as follows:
+
+```
+type DistributedExecutionOptimizer struct {
+	Endpoints api.RemoteEndpoints
+}
+
+type RemoteEndpoints interface {
+	Engines() []RemoteEngine
+}
+
+type RemoteEngine interface {
+	NewInstantQuery(opts *promql.QueryOpts, qs string, ts time.Time) (promql.Query, error)
+	NewRangeQuery(opts *promql.QueryOpts, qs string, start, end time.Time, interval time.Duration) (promql.Query, error)
+}
+```
+
+The implementation of the `RemoteEngine` will be provided by Thanos itself and will use the gRPC Query API added in [https://github.com/thanos-io/thanos/pull/5250](https://github.com/thanos-io/thanos/pull/5250).
+
+Keeping PromQL execution in Query components allows for deduplication between Prometheus pairs to happen before series are aggregated.    
+
+<img src="../img/distributed-execution-proposal-1.png" alt="Distributed query execution" width="400"/>
+
+The initial version of the solution can be found here: https://github.com/thanos-community/promql-engine/pull/139 
+
+### Query rewrite algorithm
+
+As described in the section above, the query will be rewritten using a logical optimizer into a form that is suitable for distributed execution.
+
+The proposed algorithm is as follows:
+* Start AST traversal from the bottom up.
+* If both the current node and its parent can be distributed, move up to the parent.
+* If the current node can be distributed and its parent cannot, rewrite the current node into its distributed form.
+* If the current node cannot be distributed, stop traversal.
+
+With this algorithm we try to distribute as much of the PromQL query as possible. Furthermore, even queries without aggregations, like `rate(http_requests_total[2m])`, will be rewritten into 
+```
+concat(
+  rate(http_requests_total[2m]),
+  rate(http_requests_total[2m])
+)
+```
+
+Since PromQL queries are limited in the number of steps they can evaluate, with this algorithm we achieve downsampling at query time since only a small number of samples will be sent from local Queriers to the central one.
+
+### Time-based overlap resolution
+
+Thanos stores usually have a small overlap with ingestion components (Prometheus/Receiver) due to eventually consistency from uploading and downloading TSDB blocks. As a result, the central aggregation needs a way to deduplicate samples between ingestion and storage components. 
+
+The proposed way to do time-based deduplication is by removing identical samples in the `coalesce` operator in the Thanos Engine itself. In order for data from independent Queriers to not get deduplicated, aggregations happening in remote engines must always preserve external labels from TSDB blocks that are being queried. 
```

Comment:
Oh, I see - so even with a distributed `sum()`, the external labels would be there and we would be able to deduplicate this way :muscle:

Reply:
In the Sidecar approach, the samples wouldn't be identical. Other solutions such as Cortex have a "HA Tracker" - they choose one replica from a set of replicas and only ingest samples coming from that replica. Perhaps a "remote engine" could have the option to take a similar approach. In Thanos, we could detect that a node is a Sidecar and apply this deduplication approach instead. What do you think?

Reply:
I've thought about this, and I wonder if samples not being identical is really an issue. 

The idea is to always query TSDB replicas using a remote engine. The central querier will only talk to other (remote) thanos queriers, never to stores. Remote queriers will talk to store components and deduplicate series before executing PromQL. If TSDBs pairs are found both in Prometheus and in object store, I would expect the resulting PromQL samples should be the same, and the central engine will deduplicate them. 

But maybe this is something we need to double check.